### PR TITLE
Add support for few missing pipeline and sampler vk structs

### PR DESCRIPTION
**Pull request by hakzsam**
This PR adds support for the following missing vk structs:

- VkPipelineColorWriteCreateInfoEXT
- VkPipelineRasterizationProvokingVertexStateCreateInfoEXT
- VkSamplerCustomBorderColorCreateInfoEXT
- VkSamplerReductionModeCreateInfo

### CODE REVIEW

File: fossilize.cpp [hunk 1]
```diff
@@ -3493,6 +3559,59 @@ bool StateReplayer::Impl::parse_pipeline_rendering_info(const Value &state,
...
+	return true;
+}
+
+bool StateReplayer::Impl::parse_sampler_custom_border_color(const Value &state,
+							    VkSamplerCustomBorderColorCreateInfoEXT **out_info)
+{
+	auto *info = allocator.allocate_cleared<VkSamplerCustomBorderColorCreateInfoEXT>();
+	*out_info = info;
+
+	info->customBorderColor.uint32[0] = state["customBorderColor"]["0"].GetUint();
```
Comment (HansKristian-Work) [inline fossilize.cpp:3595]: This should be an array in JSON.

File: fossilize.cpp [hunk 2]
```diff
@@ -6291,6 +6506,64 @@ static bool json_value(const VkPipelineRenderingCreateInfoKHR &create_info, Allo
...
+}
+
+template <typename Allocator>
+static bool json_value(const VkSamplerCustomBorderColorCreateInfoEXT &create_info, Allocator &alloc, Value *out_value)
+{
+	Value value(kObjectType);
+	value.AddMember("sType", create_info.sType, alloc);
+
+	Value customBorderColor(kObjectType);
+	customBorderColor.AddMember("0", create_info.customBorderColor.uint32[0], alloc);
```
Comment (HansKristian-Work) [inline fossilize.cpp:6545]: Should be kArrayType.

File: fossilize.cpp [hunk 3]
```diff
@@ -4701,10 +4908,12 @@ bool StateRecorder::Impl::copy_shader_module(const VkShaderModuleCreateInfo *cre
 bool StateRecorder::Impl::copy_sampler(const VkSamplerCreateInfo *create_info, ScratchAllocator &alloc,
                                        VkSamplerCreateInfo **out_create_info)
 {
-	if (create_info->pNext)
+	auto *info = copy(create_info, 1, alloc);
+
+	if (!copy_pnext_chain(info->pNext, alloc, &info->pNext))
```
Comment (HansKristian-Work) [inline fossilize.cpp:4913]: I think we're missing pNext hashing of samplers?

File: fossilize.cpp [hunk 4]
```diff
@@ -6291,6 +6506,64 @@ static bool json_value(const VkPipelineRenderingCreateInfoKHR &create_info, Allo
 	return true;
 }
 
+template <typename Allocator>
+static bool json_value(const VkPipelineColorWriteCreateInfoEXT &create_info, Allocator &alloc, Value *out_value)
```
Comment (HansKristian-Work) [inline fossilize.cpp:6510]: One thing about ColorWriteCreateInfo in particular is that it is affected by dynamic state, e.g. VK_DYNAMIC_STATE_COLOR_WRITE_ENABLE_BIT. I think we have to ignore color write enables array when hashing.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Overall seems okay, mostly missing roundtrip tests for these.

File: fossilize.cpp (hunk 4 above)
Comment (hakzsam) [reply fossilize.cpp:6510]: Yeah, I think the main problem is that we can't access the dynamic_info struct from there currently?

File: fossilize.cpp (hunk 1 above)
Comment (hakzsam) [reply fossilize.cpp]: Fixed.

File: fossilize.cpp (hunk 3 above)
Comment (hakzsam) [reply fossilize.cpp:4926]: Should be fixed now.

File: fossilize.cpp (hunk 2 above)
Comment (hakzsam) [reply fossilize.cpp]: Fixed.

File: fossilize.cpp [hunk 5]
```diff
@@ -4363,11 +4576,18 @@ void StateRecorder::Impl::pump_synchronized_recording(StateRecorder *recorder)
 bool StateRecorder::record_sampler(VkSampler sampler, const VkSamplerCreateInfo &create_info, Hash custom_hash)
 {
 	{
-		if (create_info.pNext)
+		// Only custom border color and reduction mode are supported for now.
+		auto *custom_border_color = find_pnext<VkSamplerCustomBorderColorCreateInfoEXT>(
```
Comment (HansKristian-Work) [inline fossilize.cpp:4580]: This is not necessary to check explicitly. We'll just error out in copy_pnext_chain instead.

File: fossilize.cpp (hunk 4 above)
Comment (HansKristian-Work) [reply fossilize.cpp:6523]: Just add a TODO, and I think we can let it slide for now.

Comment (hakzsam) [reply fossilize.cpp:6790]: Added a TODO.

**Review by HansKristian-Work — APPROVED** [review]
LGTM.

**Pull request merged**